### PR TITLE
Automatically restart validations when retriggering them. Async validation usage example and fixes.

### LINKED ISSUE #31: [Feature] | Async E-mail validation example
**Which kind of problem this feature would address? Please describe.**
Add in the current activity and fragment, an example of an async validation.

**Describe the solution you'd like**
It would be good to validate that the E-mail does not exists in some remote database, of course we don't any remote database, but we can just create a suspend fun that returns true or false based on the given E-mail with a 3 seconds delay so simulate a slow request.
It should be made as a use case and used by a custom validation.
It would also be good to be able to change the current built-in validations async or failfast settings.

**Describe alternatives you've considered**
None.

**Additional context**
None.

### PR DESCRIPTION
**Pull request by hrodrick**
#### ISSUE[#31]
* closes #31 

---

#### Description
* Implemented an async validation example on the E-mail field that runs a fake 4-seconds API request to check if the E-mail is available. We can see in the attached video a progress bar below the input that display the api progress in an unblocking way. This request is restarted automatically whenever the E-mail is changed and it is cancelled if another validation on that field changes. We can also see that the continue button is not enabled until the async validation is successful (along with the other fields' validations).
* Fixed an issue that was causing the async validations to not being cancelled nor restarted when they were being triggered again. For this, we use `Job`s in the `FlowField`'s triggerValidation functions that are inmediatelly cancelled with a customException when called again.
* Regarding the previous point, the logic in the `DefaultFieldValidationBehavior` for `FlowField`s was simplified and divided into various functions in order to be easier to understand and manage. Also added `yield()` calls for cooperative coroutine cancellation.
* Added additional documentation to illustrate this behavior.
* Added `failFast` and `async` to every built-in validation so the user can decide how those validations will run and behave.

#### Preview

https://user-images.githubusercontent.com/26490822/188683478-9836ecf0-2610-4706-b382-af544943dd27.mov

### CODE REVIEW

File: ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/SignUpFormActivity.kt [hunk 1]
```diff
@@ -76,10 +79,15 @@ class SignUpFormActivity : AppCompatActivity() {
     }
 
     private fun onEmailStatusChange(status: FieldStatus) {
+        binding.emailAvailableText.visibility = View.GONE
+        binding.emailLoadingProgressBar.visibility = View.GONE
+        binding.emailInputLayout.error = null
```
Comment (sfiamene) [inline ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/SignUpFormActivity.kt:84]: maybe `binding?.apply { ... }` can be used here?

**Review by sfiamene — APPROVED** [review]

**Review by LetoFranco — APPROVED** [review]

File: ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/EmailDoesNotExistInRemoteStorage.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    override suspend fun validate(): ValidationResult {
+        val email = emailProvider() ?: return ValidationResult(EMAIL_ALREADY_EXIST)
+        return when (verifyEmailInRemoteStorage(email)) {
+            is EmailExistenceResponse.EmailExist -> ValidationResult(EMAIL_ALREADY_EXIST)
+            is EmailExistenceResponse.EmailDoesNotExist -> ValidationResult.Correct
+        }
+    }
+
+    companion object ResultCode {
+        const val EMAIL_ALREADY_EXIST = "email-already-exist"
```
Comment (marcossalto) [inline ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/EmailDoesNotExistInRemoteStorage.kt:21]: Perhaps, we could use `EMAIL_ALREADY_EXISTS` and `email-already-exists` with S/s at the end

File: ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/SignUpFormActivity.kt [hunk 3]
```diff
@@ -76,10 +79,15 @@ class SignUpFormActivity : AppCompatActivity() {
     }
 
     private fun onEmailStatusChange(status: FieldStatus) {
+        binding.emailAvailableText.visibility = View.GONE
```
Comment (marcossalto) [inline ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/SignUpFormActivity.kt:82]: Maybe we could use `binding.apply { ... }` in order to simplify the code

**Review by marcossalto — APPROVED** [review]
Well done!
I left a couple of minor comments

File: ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/SignUpFormActivity.kt (hunk 1 above)
Comment (hrodrick) [reply ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/SignUpFormActivity.kt:84]: Yes! Nice catch

File: ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/EmailDoesNotExistInRemoteStorage.kt (hunk 2 above)
Comment (hrodrick) [reply ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/EmailDoesNotExistInRemoteStorage.kt:21]: I'm always struggling between exist and exists 😅 Thank you!

File: ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/SignUpFormActivity.kt (hunk 3 above)
Comment (hrodrick) [reply ExampleApp Android/src/main/java/com/rootstrap/flowforms/example/SignUpFormActivity.kt:82]: Yes, it fits very well here. Thanks!

**Pull request merged**